pkg/client/testing/core/fixture.go

```diff
@@ -49,201 +59,310 @@ type ObjectScheme interface {
-	gvks, _, err := o.scheme.ObjectKinds(obj)
+func (t *tracker) Update(obj runtime.Object) error {
+	return t.add(obj, true)
+}
+
+func (t *tracker) add(obj runtime.Object, replaceExisting bool) error {
+	if meta.IsListType(obj) {
+		return t.addList(obj, replaceExisting)
+	}
+
+	gvks, _, err := t.scheme.ObjectKinds(obj)
 	if err != nil {
 		return err
 	}
-	kind := gvks[0].Kind
+	if len(gvks) == 0 {
+		return fmt.Errorf("no registered kinds for %v", obj)
+	}
+	gvk := gvks[0]
+	gr := unversioned.GroupResource{Group: gvk.Group, Resource: gvk.Kind}
 
-	switch {
-	case meta.IsListType(obj):
-		if kind != "List" {
-			o.types[kind] = append(o.types[kind], obj)
-		}
+	// To avoid the object from being accidentally modified by caller
+	// after it's been added to the tracker, we always store the deep
+	// copy.
+	obj, err = t.scheme.Copy(obj)
+	if err != nil {
+		return err
+	}
+
+	if status, ok := obj.(*unversioned.Status); ok && status.Details != nil {
+		gvk.Kind = status.Details.Kind
+	}
+
+	newMeta, err := meta.Accessor(obj)
+	if err != nil {
+		return err
+	}
+
+	t.lock.Lock()
+	defer t.lock.Unlock()
 
-		list, err := meta.ExtractList(obj)
+	for i, existingObj := range t.objects[gvk] {
+		oldMeta, err := meta.Accessor(existingObj)
 		if err != nil {
 			return err
 		}
-		if errs := runtime.DecodeList(list, o.decoder); len(errs) > 0 {
-			return errs[0]
-		}
-		for _, obj := range list {
-			if err := o.Add(obj); err != nil {
-				return err
+		if oldMeta.GetNamespace() == newMeta.GetNamespace() && oldMeta.GetName() == newMeta.GetName() {
+			if replaceExisting {
+				t.objects[gvk][i] = obj
+				return nil
 			}
+			return errors.NewAlreadyExists(gr, newMeta.GetName())
+		}
+	}
+
+	if replaceExisting {
+		// Tried to update but no matching object was found.
+		return errors.NewNotFound(gr, newMeta.GetName())
+	}
+
+	t.objects[gvk] = append(t.objects[gvk], obj)
+
+	return nil
+}
+
+func (t *tracker) addList(obj runtime.Object, replaceExisting bool) error {
+	list, err := meta.ExtractList(obj)
+	if err != nil {
+		return err
+	}
+	errs := runtime.DecodeList(list, t.decoder)
+	if len(errs) > 0 {
+		return errs[0]
+	}
+	for _, obj := range list {
+		err := t.add(obj, replaceExisting)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (t *tracker) Delete(gvk unversioned.GroupVersionKind, ns, name string) error {
+	t.lock.Lock()
+	defer t.lock.Unlock()
+
+	for i, existingObj := range t.objects[gvk] {
+		objMeta, err := meta.Accessor(existingObj)
+		if err != nil {
+			return err
 		}
-	default:
-		if status, ok := obj.(*unversioned.Status); ok && status.Details != nil {
-			kind = status.Details.Kind
+		if objMeta.GetNamespace() == ns && objMeta.GetName() == name {
+			t.objects[gvk] = append(t.objects[gvk][:i], t.objects[gvk][i+1:]...)
+			return nil
 		}
-		o.types[kind] = append(o.types[kind], obj)
 	}
 
 	return nil
 }
 
+func filterByNamespaceAndName(objs []runtime.Object, ns, name string) ([]runtime.Object, error) {
```

Comment:
doc the meaning of empty here.  Empty namespace is one of the worst constructs we have.  Sometimes it is none, sometimes it is all, sometimes it is any, and sometimes it is "default".